flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/history/HistoryServer.java

```diff
@@ -0,0 +1,411 @@
+import java.nio.file.Files;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.TimerTask;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+public class HistoryServer {
+
+	private static final Logger LOG = LoggerFactory.getLogger(HistoryServer.class);
+	private static final JsonFactory jacksonFactory = new JsonFactory();
+	private static final ObjectMapper mapper = new ObjectMapper();
+
+	private static final String OVERVIEWS_FOLDER_NAME = "overviews";
+
+	private final String webAddress;
+	private final int webPort;
+	private final long webRefreshIntervalMillis;
+	private final File webDir;
+
+	private final long refreshIntervalMillis;
+	private final List<Path> refreshDirs = new ArrayList<>();
+
+	/** Map containing the JobID of all fetched jobs and the refreshDir from with they originate. */
+	private final Map<String, Path> cachedArchives = new ConcurrentHashMap<>();
+
+	private final SSLContext serverSSLContext;
+	private NettySetup netty;
+	private final Object startupShutdownLock = new Object();
+
+	private final ScheduledExecutorService executor = Executors.newSingleThreadScheduledExecutor(
+		new ExecutorThreadFactory("Flink-HistoryServer-FileFetcher"));
+	private final JobArchiveFetcherTask fetcherTask;
+
+	public static void main(String[] args) throws Exception {
+		ParameterTool pt = ParameterTool.fromArgs(args);
+		String configDir = pt.getRequired("configDir");
+
+		LOG.info("Loading configuration from {}", configDir);
+		Configuration flinkConfig = GlobalConfiguration.loadConfiguration(configDir);
+
+		HistoryServer hs = new HistoryServer(flinkConfig);
+		hs.run();
+	}
+
+	public HistoryServer(Configuration config) throws IOException {
+		if (config.getBoolean(HistoryServerOptions.HISTORY_SERVER_WEB_SSL_ENABLED) && SSLUtils.getSSLEnabled(config)) {
+			LOG.info("Enabling ssl for the history server.");
+			try {
+				this.serverSSLContext = SSLUtils.createSSLServerContext(config);
+			} catch (Exception e) {
+				throw new IOException("Failed to initialize SSLContext for the history server.", e);
+			}
+		} else {
+			this.serverSSLContext = null;
+		}
+
+		webAddress = config.getString(HistoryServerOptions.HISTRY_SERVER_WEB_ADDRESS);
+		webPort = config.getInteger(HistoryServerOptions.HISTORY_SERVER_WEB_PORT);
+		webRefreshIntervalMillis = config.getLong(HistoryServerOptions.HISTORY_SERVER_WEB_REFRESH_INTERVAL);
+
+		String webDirectory = config.getString(HistoryServerOptions.HISTORY_SERVER_WEB_DIR);
+		if (webDirectory == null) {
+			webDirectory = System.getProperty("java.io.tmpdir") + "flink-web-history-" + UUID.randomUUID();
+		}
+		webDir = new File(webDirectory);
+
+		String refreshDirectories = config.getString(HistoryServerOptions.HISTORY_SERVER_DIR);
```

Comment:
I don't know, sounds a bit inconsistent. The HS configuration would be different depending on whether you have 1 or multiple JM. With 1 JM only setting ARCHIVE_DIR is enough; for multiple HISTORY_SERVER_DIRS **must** be set.